Revise document for handling exceptions (#932)
scalar-labs/scalardb: docs/api-guide.md

    int retryCount = 0;
    TransactionException lastException = null;

    while (true) {
      if (retryCount++ > 0) {
        // Retry the transaction three times maximum in this sample code
        if (retryCount >= 3) {
          // Throw the last exception if the number of retries exceeds the maximum
          throw lastException;
        }

        // Sleep 100 milliseconds before retrying the transaction in this sample code
        TimeUnit.MILLISECONDS.sleep(100);
      }

      DistributedTransaction transaction = null;
      try {
        // Begin a transaction
        transaction = transactionManager.begin();

        // Execute CRUD operations in the transaction
        Optional<Result> result = transaction.get(...);
        List<Result> results = transaction.scan(...);
        transaction.put(...);
        transaction.delete(...);

        // Commit the transaction
        transaction.commit();
      } catch (UnsatisfiedConditionException e) {
        // You need to handle `UnsatisfiedConditionException` only if a mutation operation specifies
        // a condition. This exception indicates the condition for the mutation operation is not met

        try {

>           transaction.rollback();

Reviewer: [trivial] I guess those who read the error handling for `rollback` might wonder why just logging is okay. Maybe describing the reason that the failures of `rollback` can be ignored would be helpful for users.

Author: Thank you. I added some more explanation in ad4828e21cf86f30b1eae84b30aed1c8eb0653d8. Please take a look when you have time!

If you catch `CrudException`, it indicates that the transaction CRUD operation fails due to transient or nontransient faults. You can try retrying the transaction from the beginning, but the transaction may still fail if the cause is nontranient.
If you catch `CrudConflictException`, it indicates that the transaction CRUD operation fails due to transient faults (e.g., a conflict error). You can retry the transaction from the beginning. 

The APIs for mutation operations (`put()`/`delete()`/`mutate()`) could also throw `UnsatisfiedConditionException`.
If you can this exception, it indicates that the condition for the mutation operation is not met.
You can handle this exception according to your application requirements.

Also, the `commit()` API could throw `CommitException`, `CommitConflictException`, and `UnknownTransactionStatusException`.
If you catch `CommitException`, it indicates that committing the transaction fails due to transient or nontransient faults. You can try retrying the transaction from the beginning, but the transaction may still fail if the cause is nontranient.

> If you catch `CommitConflictException`, it indicates that committing the transaction fails due to transient faults (e.g., a conflict error). You can retry the transaction from the beginning.

Reviewer 2: ```suggestion
If you catch `CommitConflictException`, this indicates that committing the transaction has failed due to transient faults (e.g., a conflict error). You can retry the transaction from the beginning.
```

> If you catch `UnknownTransactionStatusException`, it indicates that you are not sure if the transaction succeeds or not.

Reviewer 2: ```suggestion
If you catch `UnknownTransactionStatusException`, this indicates that you are not sure if the transaction succeeds or not.
```

Although I edited the above sentence a bit, it is a bit unclear to me. Should `this indicates that you are not sure if the transaction succeeds or not` be something like `this indicates that the status of the transaction, whether it has succeeded or not, is unknown` instead?

Author: Sounds good. Thank you! Fixed in 0b33081cfbc4f37516948c8d6c4b84821c2f0a29.

Although not illustrated in the sample code, the `resume()` API could also throw a `TransactionNotFoundException`.
This exception indicates that the transaction associated with the specified ID was not found, and it might have been expired.
In such cases, you can retry the transaction from the beginning since the cause of this exception is basically transient.

In the sample code, for `UnknownTransactionStatusException`, the transaction doesn't retry because the cause of the exception is nontransient.
Also, for `UnsatisfiedConditionException`, the transaction doesn't retry because how to handle this exception depends on your application requirements.
For other exceptions, the transaction tries retrying because the cause of the exception is transient or nontransient.
If the cause of the exception is transient, the transaction may succeed if you retry it.
However, if the cause of the exception is nontransient, the transaction may still fail even if you retry it.
In such a case, you will exhaust the number of retries.

Please note that if you begin a transaction by specifying a transaction ID, you must use a different ID when you retry the transaction.

> And, in the sample code, the transaction retries three times maximum and sleeps fixed 100 milliseconds before it retries.

Reviewer 2: ```suggestion
And, in the sample code, the transaction retries three times maximum and sleeps for 100 milliseconds before it retries.
```

> But you can choose a retry policy according to your application requirements, such as exponential backoff.

Reviewer 2: ```suggestion
But you can choose a retry policy, such as exponential backoff, according to your application requirements.
```
